modules/core/src/effects/mask/mask-effect.js

```diff
@@ -0,0 +1,146 @@
+import {
+  log,
+  Texture2D,
+  ProgramManager
+  // readPixelsToArray
+} from '@luma.gl/core';
+// import {SolidPolygonLayer} from '@deck.gl/layers';
+import MaskPass from '../../passes/mask-pass';
+import Effect from '../../lib/effect';
+import {default as mask} from '../../shaderlib/mask/mask';
+import {getMaskProjectionMatrix, getMaskViewport} from './utils';
+
+// Class to manage mask effect
+export default class MaskEffect extends Effect {
+  constructor(props) {
+    super(props);
+    this.programManager = null;
+    this.dummyMaskMap = null;
+    this.mask = false;
+    this.enableForPicking = true;
+  }
+
+  preRender(gl, {layers, layerFilter, viewports, onViewportActive, views}) {
+    const maskIds = new Set(
+      layers.map(({props}) => props.maskId).filter(maskId => maskId !== undefined)
+    );
+    if (maskIds.size === 0) {
+      return;
+    }
+    log.assert(maskIds.size === 1, 'Only one mask layer supported, but multiple maskIds specified');
+
+    this.mask = true;
+    const maskId = [...maskIds][0];
+
+    const maskLayer = this.getMaskLayer(maskId, layers);
+    if (!this.maskPass) {
+      this.maskPass = new MaskPass(gl);
+      this.maskMap = this.maskPass.maskMap;
+    }
+    const {maskPass, maskMap} = this;
+    if (!this.programManager) {
+      this.programManager = ProgramManager.getDefaultProgramManager(gl);
+      this.programManager.addDefaultModule(mask);
+    }
+
+    if (!this.dummyMaskMap) {
+      this.dummyMaskMap = new Texture2D(gl, {
+        width: 1,
+        height: 1
+      });
+    }
+
+    // Using the 'positions' attribute will work for the PolygonLayer,
+    // but not for all layers
+    const {positions} = maskLayer.state.attributeManager.attributes;
+    const layerViewport = viewports[0];
+    const maskViewport = getMaskViewport(positions, layerViewport, maskMap);
+    this.maskProjectionMatrix = getMaskProjectionMatrix(maskViewport);
+
+    maskPass.render({
```

Comment:
Skip rerendering the mask framebuffer if using the same viewport and the mask layer has not changed.

Example: if you move your mouse across the map, picking is performed on every animation frame. There is no reason to rerender the mask.